crosscluster/logical: hydrate table descriptors with types for event decoding

msbutler opened the pull request:
This hydrates the source side table descriptors with their referencing type
descriptors if they have any. This allows ldr ingestion to properly decode
tables with user defined types.

Note that LDR will still fast fail on replicating a table with user defined
types (unless the SKIP SCHEMA CHECK OPTION is used) because we still need to
teach LDR to:
- validate that the source and destination enum's are _physically_ identical
  (contain the same map from int to enum values)
- support UDT replication on sql path, which currently fails because the passed
  in datum points to the source side udt instead of the destination udt.

As another step down the line, we should relax the requirement that UDTs must
physically equivalent. Instead, they only need to be _logically_ equivalent
(i.e. have the same user facing enum values). Supporting this requires
remapping the ints from two logically equivalent enums.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131130)
<!-- Reviewable:end -->

msbutler committed:
importer: make import type resolver public

Epic: none

Release note: none

msbutler committed:
crosscluster/producer: grab type descriptors during LDR physical planning

Epic: none

Release note: none

msbutler committed:
crosscluster/logical: hydrate table descriptors with types for event decoding

This hydrates the source side table descriptors with their referencing type
descriptors if they have any. This allows ldr ingestion to properly decode
tables with user defined types.

Note that LDR will still fast fail on replicating a table with user defined
types (unless the SKIP SCHEMA CHECK OPTION is used) because we still need to
teach LDR to:
- validate that the source and destination enum's are _physically_ identical
  (contain the same map from int to enum values)
- support UDT replication on sql path, which currently fails because the passed
  in datum points to the source side udt instead of the destination udt.

As another step down the line, we should relax the requirement that UDTs must
physically equivalent. Instead, they only need to be _logically_ equivalent
(i.e. have the same user facing enum values). Supporting this requires
remapping the ints from two logically equivalent enums.

Epic: none

Release note: none

msbutler commented on pkg/sql/importer/import_type_resolver.go:
```diff
@@ -52,7 +52,7 @@ func makeImportTypeResolver(typeDescs []*descpb.TypeDescriptor) importTypeResolv
 // different schemas), this implementation will return a "feature unsupported"
 // error.
-func (i importTypeResolver) ResolveType(
+func (i ImportTypeResolver) ResolveType(
```
I'm wondering if there's a better `TypeResolver` is should use?

msbutler replied:
This type resolver is insufficient for the sql write path. Ideally the type resolver used by LDR can take in foreign types with mismatched OIDs and allow the sql path to just work. Currently the sql path writes panic because we hydrate the datums ingested using the source side UDTs, which have a different OID than on the destination side.

msbutler commented on pkg/ccl/crosscluster/producer/replication_manager.go:
```diff
@@ -137,20 +141,52 @@ func (r *replicationStreamManagerImpl) PlanLogicalReplication(
+		}
+		typeIDs, _, err := td.GetAllReferencedTypeIDs(dbDesc,
+			func(id descpb.ID) (catalog.TypeDescriptor, error) {
+				if id > 0 {
```
what's the deal with this function passed to `GetAllREferencedTypeIDs`? as the panic suggests, it doesn't seem to get called.

msbutler commented on pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go:
```diff
@@ -341,10 +342,17 @@ func (p *partitionedStreamClient) PlanLogicalReplication(
 		descMap[int32(desc.ID)] = desc
 	}
 
+	// TODO(msbutler): unclear if we should persist the types as pointers in the protos.
```
ignore this comment. i like how i stored the types in `stream.proto`

fqazi reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131130)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, and @rafiss)

___
*[`pkg/ccl/crosscluster/logical/logical_replication_job.go` line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cC_03AHsAJlK4Kj97:-O8cC_03AHsAJlK4Kj98:b-ggrvyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/afcc0fc5358e3d2af17709da4eb69ab54e261aa8/pkg/ccl/crosscluster/logical/logical_replication_job.go#L363)):*
> ```Go
> 
> 	// TODO(msbutler): is this import type resolver kosher? Should put in a new package.
> 	importResolver := importer.MakeImportTypeResolver(plan.SourceTypes)
> ```

Probably want to in a separate PR move this out / rename this later.
___
*[`pkg/ccl/crosscluster/producer/replication_manager.go` line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131130#gh-1788252139:-O8cC8v39H5lBS1Ytz2v:b-qbutbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad184be84a6e26676e873d6d54a9d995097b6b6a/pkg/ccl/crosscluster/producer/replication_manager.go#L163)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

what's the deal with this function passed to `GetAllREferencedTypeIDs`? as the panic suggests, it doesn't seem to get called.
</blockquote></details>

This code is only focused on default / update expressions? Have your tests included that. i.e. Have a cast in your default expression.
___
*[`pkg/ccl/crosscluster/producer/replication_manager.go` line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cCRaG5d9eE7xxXPF_:-O8cCRaG5d9eE7xxXPFa:bale275) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad184be84a6e26676e873d6d54a9d995097b6b6a/pkg/ccl/crosscluster/producer/replication_manager.go#L170)):*
> ```Go
> 				}
> 				foundTypeDescriptors[id] = struct{}{}
> 				return descriptors.ByIDWithoutLeased(r.txn.KV()).WithoutNonPublic().Get().Type(ctx, id)
> ```

Why do we avoid leasing here?
___
*[`pkg/sql/importer/import_type_resolver.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131130#gh-1788251406:-O8cD7um0vFlfZC0oRfu:bp6at5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a0f7be2955d95e16e09736cb93c5dcc877f5cf/pkg/sql/importer/import_type_resolver.go#L55)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

This type resolver is insufficient for the sql write path. Ideally the type resolver used by LDR can take in foreign types with mismatched OIDs and allow the sql path to just work. Currently the sql path writes panic because we hydrate the datums ingested using the source side UDTs, which have a different OID than on the destination side. I can write up a separate issue to address this
</blockquote></details>

I think this is probably the better option. We can probably make this more generic implementation more generic. i.e. Extend it so that it can rewrite OIDs upon hydration as a feature, which only LDR needs.

Or implement a LDR specific version on top with that logic.
___
*[`pkg/ccl/crosscluster/logical/logical_replication_job_test.go` line 1974 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cCsWW9k20Rru2XVuX:-O8cCsWW9k20Rru2XVuY:bj0vqdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/afcc0fc5358e3d2af17709da4eb69ab54e261aa8/pkg/ccl/crosscluster/logical/logical_replication_job_test.go#L1974)):*
> ```Go
> 	dbB.Exec(t, "CREATE TYPE my_enum AS ENUM ('one', 'two', 'three')")
> 
> 	dbA.Exec(t, "CREATE TABLE data (pk INT PRIMARY KEY, val my_enum)")
> ```

To hit some of the code paths above we can have fun and add default expressions which use the type.


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131130)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @rafiss)

___
*[`pkg/ccl/crosscluster/logical/logical_replication_job.go` line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cC_03AHsAJlK4Kj97:-O8cEsu00pQVOOdyTnhU:bnxb3b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/afcc0fc5358e3d2af17709da4eb69ab54e261aa8/pkg/ccl/crosscluster/logical/logical_replication_job.go#L363)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Probably want to in a separate PR move this out / rename this later.
</blockquote></details>

hrm, are you ok that I make the ImportTypeResolver public temporarily in this PR. Once we have the LDR specific type resolver, we can make it private again.
___
*[`pkg/ccl/crosscluster/producer/replication_manager.go` line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131130#gh-1788252139:-O8g6zxA7-WPi08AIjBG:b-kkulgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad184be84a6e26676e873d6d54a9d995097b6b6a/pkg/ccl/crosscluster/producer/replication_manager.go#L163)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This code is only focused on default / update expressions? Have your tests included that. i.e. Have a cast in your default expression.
</blockquote></details>

ah yes, i hit this panic after i added a default expression in my unit test. I feel like this `GetAllReferencedTypeIDs` function could be better documented to explain that the `getType` function only gets called when a column with the UDT in that table has a default expression. 
___
*[`pkg/ccl/crosscluster/producer/replication_manager.go` line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cCRaG5d9eE7xxXPF_:-O8g7QNZ2OSmfAdup2W5:bj4ze0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad184be84a6e26676e873d6d54a9d995097b6b6a/pkg/ccl/crosscluster/producer/replication_manager.go#L170)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Why do we avoid leasing here?
</blockquote></details>

This was copy pasta. We grab the descriptor with the lease now.
___
*[`pkg/ccl/crosscluster/logical/logical_replication_job_test.go` line 1974 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cCsWW9k20Rru2XVuX:-O8g7aBl5k7Ea_vZWoFc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/afcc0fc5358e3d2af17709da4eb69ab54e261aa8/pkg/ccl/crosscluster/logical/logical_replication_job_test.go#L1974)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

To hit some of the code paths above we can have fun and add default expressions which use the type.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131130)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @rafiss)

___
*[`pkg/sql/importer/import_type_resolver.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131130#gh-1788251406:-O8gAuyg1F1wOutJT6er:b2c2ux6) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a0f7be2955d95e16e09736cb93c5dcc877f5cf/pkg/sql/importer/import_type_resolver.go#L55)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think this is probably the better option. We can probably make this more generic implementation more generic. i.e. Extend it so that it can rewrite OIDs upon hydration as a feature, which only LDR needs.

Or implement a LDR specific version on top with that logic.
</blockquote></details>

wrote up this issue and assigned it to you https://github.com/cockroachdb/cockroach/issues/132164


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r4, 3 of 3 files at r5, 3 of 4 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-:-O8gGPggDYAiEDFrIBxe:b-8ewgme)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt and @rafiss)

___
*[`pkg/ccl/crosscluster/logical/logical_replication_job.go` line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8cC_03AHsAJlK4Kj97:-O8gFxpUCrrf33CoECX_:b-28u7yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/afcc0fc5358e3d2af17709da4eb69ab54e261aa8/pkg/ccl/crosscluster/logical/logical_replication_job.go#L363)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

hrm, are you ok that I make the ImportTypeResolver public temporarily in this PR. Once we have the LDR specific type resolver, we can make it private again.
</blockquote></details>

Done.

Yeah, I'm good with making this public. And then come up with a clean interface later.


<!-- Sent from Reviewable.io -->

msbutler committed:
crosscluster/logical: get type descriptors during planning

This will allow us to do type descriptor validation when we check for
replicating table equivalency.

Epic: none

Release note: none

fqazi reviewed (approved):
> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131130)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @msbutler, and @rafiss)

___
*[`pkg/ccl/crosscluster/producer/replication_manager.go` line 147 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8gJLenDC8WDaeSVPwr:-O8gJLenDC8WDaeSVPws:b-cikucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d105a0f9853ae1e298e58f317820b2335dfd043/pkg/ccl/crosscluster/producer/replication_manager.go#L147)):*
> ```Go
> ) ([]descpb.TypeDescriptor, map[descpb.ID]struct{}, error) {
> 	descriptors := txn.Descriptors()
> 	dbDesc, err := descriptors.MutableByID(txn.KV()).Database(ctx, td.GetParentID())
> ```

We can just get a leased one for the database descriptor, since I don't think these get modified.
___
*[`pkg/ccl/crosscluster/producer/replication_manager.go` line 165 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131130#-O8gK0WS31dIrVReRFjU:-O8gK0WS31dIrVReRFjV:bi3pbnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d105a0f9853ae1e298e58f317820b2335dfd043/pkg/ccl/crosscluster/producer/replication_manager.go#L165)):*
> ```Go
> 		foundTypeDescriptors[typeID] = struct{}{}
> 
> 		typeDesc, err := descriptors.MutableByID(txn.KV()).Type(ctx, typeID)
> ```

If we don't intend to write these back out and just want a modifiable copy, we can use a leased descriptor and the builder API. Just something to consider later, since it will avoid an unnecessary KV roundtrip to get the descriptor.


<!-- Sent from Reviewable.io -->

msbutler commented:
TFTR!

bors=fqazi

msbutler commented:
bors r+

rafiss commented on pkg/ccl/crosscluster/logical/logical_replication_job_test.go:
```diff
@@ -1942,6 +1942,52 @@ func TestLogicalReplicationSchemaChanges(t *testing.T) {
+	dbB.Exec(t, "INSERT INTO data VALUES (2)")
+
+	var jobAID jobspb.JobID
+	dbA.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE data ON $1 INTO TABLE data with skip schema check", dbBURL.String()).Scan(&jobAID)
```
should we remove the check from here: https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/catalog/tabledesc/logical_replication_helpers.go#L163-L168

that way we don't need the `skip schema check`

msbutler replied:
i think we should remove it once we fix this issue https://github.com/cockroachdb/cockroach/issues/132167
